### PR TITLE
Imu fix

### PR DESCRIPTION
**Pull request by jzheng84**
This PR reorganizes the IMU code for the FSM-9.

Major changes:
* Launch parameter for specifying the correction matrix for putting the IMU in the same coordinate frame as the robot
* Read and compare the magnetometer heading to the filtered heading, and print out a warning when there is a deviation
* Clean up code and address meeting comments

### CODE REVIEW

File: igvc_platform/launch/imu.launch [hunk 1]
```diff
@@ -1,21 +1,9 @@
...
-      <!-- 3.09938308897 -->
-      <!-- 2.93163273405 -->
-      <!-- 2.73621983653 -->
-      <!-- at comp -->
-      <!-- 1.0726804111 -->
-      <!-- 1.36263847313 -->
-    </node>
+    <node name="imu" pkg="igvc_platform" type="imu" output="screen" />
+    	<!-- Matrix to put IMU in same coordinate system as the robot. -->
+	    <rosparam param="imu_orientation_correction">[1, 0, 0, 0, -1, 0, 0, 0, -1]</rosparam>
```
Comment (JasonGibson274) [inline igvc_platform/launch/imu.launch]: I would line this up as a matrix, just for quicker scanning

File: igvc_platform/src/imu/main.cpp [hunk 2]
```diff
@@ -71,32 +62,58 @@ int main(int argc, char** argv)
...
+
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "imu");
+
+  ros::NodeHandle nh;
+  ros::NodeHandle pNh("~");
+
+  // Get matrix to correct IMU orientation.
+  std::vector<double> imu_orientation_correction;
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: we do not check if param exists and warns. There is a header in utils that will handle this inline.

File: igvc_platform/src/imu/main.cpp [hunk 3]
```diff
@@ -71,32 +62,58 @@ int main(int argc, char** argv)
...
+  ros::Publisher imu_raw_pub = nh.advertise<sensor_msgs::Imu>("/imu_raw", 1000);
+
+  // sequence of published messages - should be monotonically increasing.
+  int seq = 0;
+
+  // Initialize IMU.
+  FreespaceDeviceId ids[1];
+  int ret = initializeIMU(ids);
+  if (ret != FREESPACE_SUCCESS) {
+    return 0;
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: I would print an error here instead of quietly ending.

File: igvc_platform/src/imu/main.cpp [hunk 4]
```diff
@@ -71,32 +62,58 @@ int main(int argc, char** argv)
...
 
+  // Variable to store data that is read from the IMU.
+  freespace_message response;
+
   while (ros::ok())
   {
     ros::spinOnce();
 
+    // Read the IMU message every loop. Timeout after 1000 milliseconds.
     ret = freespace_readMessage(ids[0], &response, 1000);
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: @ someone else. Is it clear this is blocking?

File: igvc_platform/src/imu/main.cpp [hunk 5]
```diff
@@ -115,128 +132,136 @@ int main(int argc, char** argv)
       ROS_ERROR_STREAM("error got unknown message type " << response.messageType);
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: can we preface with "IMU:"

File: igvc_platform/src/imu/main.cpp [hunk 6]
```diff
@@ -115,128 +132,136 @@ int main(int argc, char** argv)
...
     MultiAxisSensor orientation_msg;
     ret = freespace_util_getAngPos(&response.motionEngineOutput, &orientation_msg);
     if (ret != FREESPACE_SUCCESS)
     {
       ROS_ERROR_STREAM("failed to read angular position " << ret);
     }
+    tf::Quaternion quaternion_raw(orientation_msg.x, orientation_msg.y, orientation_msg.z, orientation_msg.w);
+    double roll, pitch, yaw;
+    tf::Matrix3x3(quaternion_raw).getRPY(roll, pitch, yaw);
+    ROS_INFO("Filtered heading: %f", yaw);
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: maybe not print out to terminal. We have a launch file that will convert quat to r,p,y on the fly on util

File: igvc_platform/src/imu/main.cpp [hunk 7]
```diff
@@ -115,128 +132,136 @@ int main(int argc, char** argv)
...
-    orientation.w = rotated.w();
-
+    //orientation.x = quaternion_raw.x();
+    //orientation.y = quaternion_raw.y();
+    //orientation.z = quaternion_raw.z();
+    //orientation.w = quaternion_raw.w();
+    // This assumes flat ground and uses magnetometer for absolute heading.
+    // TODO: This assumption may not hold when we switch to 3D lidar, and stop using flat ground assumption
+    tf::Quaternion quaternion_mag;
+    quaternion_mag.setRPY(0,0,yaw_mag);
```
Comment (JasonGibson274) [inline igvc_platform/src/imu/main.cpp]: We should stop doing this.

File: igvc_platform/launch/imu.launch (hunk 1 above)
Comment (jzheng84) [reply igvc_platform/launch/imu.launch]: Done

File: igvc_platform/src/imu/main.cpp (hunk 2 above)
Comment (jzheng84) [reply igvc_platform/src/imu/main.cpp]: Done

File: igvc_platform/src/imu/main.cpp (hunk 3 above)
Comment (jzheng84) [reply igvc_platform/src/imu/main.cpp]: Done

File: igvc_platform/src/imu/main.cpp (hunk 5 above)
Comment (jzheng84) [reply igvc_platform/src/imu/main.cpp]: Done

File: igvc_platform/src/imu/main.cpp (hunk 6 above)
Comment (jzheng84) [reply igvc_platform/src/imu/main.cpp]: Done

File: igvc_platform/src/imu/main.cpp (hunk 7 above)
Comment (jzheng84) [reply igvc_platform/src/imu/main.cpp]: Will address when we start switching code to handle 3D cases

**Comment by oswinso** [comment]
Irrelevant with the new IMU.

**Pull request closed without merging**